platform/rct/hal/usb_stm32/USB-CDC_device_stm32.c

```diff
@@ -1,52 +1,250 @@
-#include "USB-CDC_device.h"
+#include <FreeRTOS.h>
+#include <task.h>
+#include <portmacro.h>
+#include <semphr.h>
+#include <timers.h>
+#include <queue.h>
+
+#include <usb_lib.h>
+#include <usb_desc.h>
+#include <usb_mem.h>
+#include <hw_config.h>
+#include <usb_istr.h>
+#include <usb_pwr.h>
+#include <usb_prop.h>
+
 #include <string.h>
+#include <stdbool.h>
+
+#include <USB-CDC_device.h>
+
+#define USB_BUF_ELTS(in, out, bufsize) ((in - out + bufsize) % bufsize)
+
+static struct {
+	uint8_t USB_Rx_Buffer[VIRTUAL_COM_PORT_DATA_SIZE];
+	uint8_t USB_Tx_Buffer[VIRTUAL_COM_PORT_DATA_SIZE];
+	volatile uint32_t USB_Tx_ptr_in;
+	volatile uint32_t USB_Tx_ptr_out;
+	volatile uint32_t USB_Rx_ptr_in;
+	volatile uint32_t USB_Rx_ptr_out;
+	volatile int _init_flag;
+	xSemaphoreHandle _rlock;
+	xSemaphoreHandle _wlock;
+} usb_state;
+
+/* ST's USB hardware doesn't include an internal pullup resistor on
+ * USB_DP for some odd reason and instead relies on an externally
+ * gated (via a set of transistors) pull up attached to a
+ * USB_DISCONNECT pin. As this is only used during startup (we don't
+ * do any dynamic re-enumeration), we can shortcut this by having a
+ * static pullup, and simply causing a glitch on the line by forcing
+ * USB_DP low before we initialize the rest of the USB hardware. It's
+ * a little hacky, but doesn't seem to cause any ill effects. */
+static int USB_CDC_force_reenumeration(void)
+{
+	GPIO_InitTypeDef GPIO_InitStructure;
+
+	RCC_AHBPeriphClockCmd(RCC_AHBPeriph_GPIOA, ENABLE);
+	GPIO_InitStructure.GPIO_Pin = GPIO_Pin_12;
+	GPIO_InitStructure.GPIO_Speed = GPIO_Speed_50MHz;
+	GPIO_InitStructure.GPIO_Mode = GPIO_Mode_OUT;
+	GPIO_InitStructure.GPIO_OType = GPIO_OType_PP;
+	GPIO_InitStructure.GPIO_PuPd = GPIO_PuPd_NOPULL;
+	GPIO_Init(GPIOA, &GPIO_InitStructure);
+
+ 	GPIO_ResetBits(GPIOA, GPIO_Pin_12);
+
+	GPIO_InitStructure.GPIO_OType = GPIO_OType_OD;
+	GPIO_Init(GPIOA, &GPIO_InitStructure);
+
+}
 
-#include <usbd_cdc_core.h>
-#include <usbd_usr.h>
-#include <usbd_conf.h>
-#include <usbd_desc.h>
-#include <usbd_cdc_vcp.h>
+/* Public API */
+int USB_CDC_device_init(const int priority)
+{
+	/* Perform a full USB reset */
+	USB_CDC_force_reenumeration();
+
+	/* Create read/write mutexes */
+	usb_state._rlock = xSemaphoreCreateMutex();
+	usb_state._wlock = xSemaphoreCreateMutex();
+
+	if (usb_state._rlock == NULL || usb_state._wlock == NULL) {
```

Comment:
NIT: Use Yoda checks.  So `NULL == usb_state._rlock`.  This way if there were an accidental single equals... compiler catches it.